platform/ATParser.h

```diff
@@ -0,0 +1,258 @@
+    * a timeout occurs.
+    *
+    * @param response scanf-like format string of response to expect
+    * @param ... all scanf-like arguments to extract from response
+    * @return true only if response is successfully matched
+    */
+    bool recv(const char *response, ...)
+#if defined(__GNUC__) || defined(__CC_ARM)
+    __attribute__ ((__format__(__scanf__, 2, 3)))
+#endif
+    ;
+    bool vrecv(const char *response, va_list args);
+
+    /**
+    * Write a single byte to the underlying stream
+    *
+    * @param c The byte to write
+    * @return The byte that was written or -1 during a timeout
+    */
+    int putc(char c);
+
+    /**
+    * Get a single byte from the underlying stream
+    *
+    * @return The byte that was read or -1 during a timeout
+    */
+    int getc();
+
+    /**
+    * Write an array of bytes to the underlying stream
+    *
+    * @param data the array of bytes to write
+    * @param size number of bytes to write
+    * @return number of bytes written or -1 on failure
+    */
+    int write(const char *data, int size);
+
+    /**
+    * Read an array of bytes from the underlying stream
+    *
+    * @param data the destination for the read bytes
+    * @param size number of bytes to read
+    * @return number of bytes read or -1 on failure
+    */
+    int read(char *data, int size);
+
+    /**
+    * Direct printf to underlying stream
+    * @see ::printf
+    *
+    * @param format format string to pass to printf
+    * @param ... arguments to printf
+    * @return number of bytes written or -1 on failure
+    */
+    int printf(const char *format, ...)
+#if defined(__GNUC__) || defined(__CC_ARM)
+    __attribute__ ((__format__(__printf__, 2, 3)))
+#endif
+    ;
+    int vprintf(const char *format, va_list args);
+
+    /**
+    * Direct scanf on underlying stream
+    * @see ::scanf
+    *
+    * @param format format string to pass to scanf
+    * @param ... arguments to scanf
+    * @return number of bytes read or -1 on failure
+    */
+    int scanf(const char *format, ...)
+#if defined(__GNUC__) || defined(__CC_ARM)
+    __attribute__ ((__format__(__scanf__, 2, 3)))
+#endif
+    ;
+    int vscanf(const char *format, va_list args);
+
+    /**
+    * Attach a callback for out-of-band data
+    * 
+    * @param prefix string on when to initiate callback
+    * @param func callback to call when string is read
+    * @note out-of-band data is only processed during a scanf call
+    */
+    void oob(const char *prefix, mbed::Callback<void()> func);
+
+    /**
+    * Attach a callback for out-of-band data
+    *
+    * @param prefix string on when to initiate callback
+    * @param obj pointer to object to call member function on
+    * @param method callback to call when string is read
+    * @note out-of-band data is only processed during a scanf call
+    */
+    template <typename T, typename M>
+    void oob(const char *prefix, T *obj, M method) {
```

Comment:
The definition above takes in a callback:
https://github.com/ARMmbed/mbed-os/pull/4119/files#diff-feeda9e18e6dcf0b25b0c69eae598e78R230

We also provide a function that converts object+member function into a callback:
``` cpp
class Thing { void doit() { printf("hi"); } };
Thing thing;

atparser.oob("AT!", callback(&thing, &Thing::doit));
```

The reason we do this is to keep the mess that is `Callback.h` from spreading.